SIMSBIOHUB-620: Manage System Alerts

mauberti-bc opened the pull request:
## Links to Jira Tickets

- https://apps.nrs.gov.bc.ca/int/jira/browse/SIMSBIOHUB-620

## Description of Changes

- Adds the ability for system admins to create Alerts that all users see
- Displays alerts on relevant pages, with each page requesting only relevant alerts (eg. alerts of type "Survey" for Survey page)
- Repurposes Manage Users page to generic Admin page with controls for managing users and alerts
- Adds system alerts to the database seeds

## Testing Notes

- Should be able to create, edit and delete alerts
- Alerts should only display on the appropriate pages (eg. General on the Survey page)

NickPhura committed:
Initial migration to add alert table WIP)

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into SIMSBIOHUB-620

mauberti-bc committed:
Merge branch 'dev' of github.com:bcgov/biohubbc into SIMSBIOHUB-620

mauberti-bc committed:
add alert endpoints, services, repos and frontend components copying funding sources patterns

mauberti-bc committed:
alerts working

mauberti-bc committed:
style alert table

mauberti-bc committed:
separate requests for active and expired alerts & rename manage users to admin page

MacDeluca committed:
Merge branch 'dev' into SIMSBIOHUB-620

mauberti-bc committed:
Merge branch 'dev' of github.com:bcgov/biohubbc into SIMSBIOHUB-620

mauberti-bc committed:
Merge branch 'SIMSBIOHUB-620' of github.com:bcgov/biohubbc into SIMSBIOHUB-620

mauberti-bc committed:
fix tests and code smells

mauberti-bc committed:
address code smell

mauberti-bc committed:
add updateAlert

NickPhura commented:
Converting to draft temporarily to save Openshift resources.

NickPhura committed:
Merge branch 'dev' into SIMSBIOHUB-620

NickPhura commented on database/src/migrations/20240927000000_alert_table.ts:
```diff
@@ -0,0 +1,127 @@
+      name               varchar(100)       NOT NULL,
+      message            varchar(250)       NOT NULL,
+      data               json,
+      severity           varchar(100)       NOT NULL,
```
nit: missing comment for this column, below.

NickPhura replied:
Also, does this need to reference a lookup table of severity codes?

mauberti-bc replied:
The severity values are just MUI severity codes. I switched to an enum
```
CREATE TYPE alert_severity AS ENUM ('info', 'warning', 'error', 'success');
```

NickPhura commented on app/src/hooks/api/useAlertApi.ts:
```diff
@@ -0,0 +1,88 @@
+ */
+export const useAlertApi = (axios: AxiosInstance) => {
+  /**
+   * Get project alert details based on its ID for viewing purposes.
```
nit: Doc seems to be out of date.

NickPhura replied:
All of the doc strings seem slightly off. They all mention 'project' even though the alerts arent project specific.

mauberti-bc replied:
Fixed

NickPhura commented on app/src/features/alert/banner/SystemAlertBanner.tsx:
```diff
@@ -0,0 +1,73 @@
+import { useEffect, useState } from 'react';
+import { TransitionGroup } from 'react-transition-group';
+
+export enum SystemAlertBannerEnum {
```
I would maybe move this into `useAlertApi.interface.ts`.

The filter param `IAlertFilterParams.types` is just defined as a `string[]`, but it looks like its value should be an array of this type.

mauberti-bc replied:
Thanks, fixed.

NickPhura commented on app/src/features/admin/users/access-requests/list/pending/AccessRequestPendingList.tsx:
```diff
@@ -33,7 +32,10 @@ interface IAccessRequestPendingListProps {
-  const { accessRequests, codes, refresh } = props;
+  const { accessRequests, refresh } = props;
+
+  const codesContext = useCodesContext();
```
Will want to add a useEffect to ensure codes are loaded:

```typescript
  useEffect(() => {
    codesContext.codesDataLoader.load();
  }, [codesContext.codesDataLoader]);
```

NickPhura replied:
(TBH, I've been wondering about just putting the useEffect in the `useCodesContext` hook, that way no one ever has to think about it. But, something for another PR.)

mauberti-bc replied:
Fixed

NickPhura commented on api/src/openapi/schemas/alert.ts:
```diff
@@ -0,0 +1,62 @@
+ */
+export const systemAlertSchema: OpenAPIV3.SchemaObject = {
+  ...baseSystemAlertSchema,
+  required: ['alert_id', 'name', 'message', 'data', 'alert_type_id', 'record_end_date', 'severity'],
```
Add the 'additionalProperties: false' line

NickPhura replied:
Could maybe just add it to the base schemas as well

NickPhura commented on api/src/openapi/schemas/alert.ts:
```diff
@@ -0,0 +1,62 @@
+ * Schema for creating system alerts
+ */
+export const systemAlertCreateSchema: OpenAPIV3.SchemaObject = {
+  ...baseSystemAlertSchema,
```
Add the 'additionalProperties: false' line

NickPhura commented on api/src/paths/alert/index.ts:
```diff
@@ -0,0 +1,244 @@
+        type: 'array',
+        items: {
+          type: 'string',
+          nullable: true
```
You probably dont want this to be nullable. Since this is inside the `items` section, this means that the array is allowed to have null values.
```
types=['type1', null, 'type2']
```
I would just remove the nullable entirely, and leave the query param as not required. That way the caller can omit `types`, or else include it with non-null values.

Same with the query params below.

mauberti-bc replied:
Fixed

NickPhura commented on api/src/paths/alert/index.ts:
```diff
@@ -0,0 +1,244 @@
+    return {
+      and: [
+        {
+          discriminator: 'SystemUser',
```
This should be `'SystemRole'`

There are a few other endpoints I notice with the same issue.

Searching `discriminator: 'SystemUser',` seems to find the other 3 I can see.

NickPhura commented on api/src/paths/alert/index.ts:
```diff
@@ -0,0 +1,244 @@
+      content: {
+        'application/json': {
+          schema: {
+            type: 'object',
+            additionalProperties: false,
+            required: ['alert_id'],
+            properties: {
+              alert_id: {
+                type: 'number'
+              }
+            }
+          }
```
(Just a general comment, no change needed) 

We have a mixture of patterns for responses in SIMS atm.  Returning the newly created record or the primary ID is fine, but its also fine to just return 200 and nothing else:

```
responses: {
    200: {
      description: 'Create alert OK'
    },
    ....
}

NickPhura commented on api/src/paths/alert/{alertId}/index.test.ts:
```diff
@@ -0,0 +1,243 @@
+      await requestHandler(mockReq, mockRes, mockNext);
+
+      expect(mockRes.jsonValue).to.eql(mockAlert);
+      expect(mockDBConnection.open).to.have.been.calledOnce;
+      expect(mockDBConnection.commit).to.have.been.calledOnce;
```
Nice 👍

NickPhura replied:
Could add a call to check `release` was called too.

(Lots of our old tests don't have these checks, but will help moving forward if we add them)

NickPhura commented on api/src/paths/alert/{alertId}/index.ts:
```diff
@@ -0,0 +1,310 @@
+    return {
+      and: [
+        {
+          discriminator: 'SystemUser',
```
Should be `SystemRole`

NickPhura commented on database/src/seeds/03_basic_project_survey_setup.ts:
```diff
@@ -770,3 +775,31 @@ const insertAccessRequest = () => `
+    )
+  VALUES (
+    (SELECT alert_type_id FROM alert_type ORDER BY random() LIMIT 1),
+    $$${faker.lorem.words(3)}$$,
```
I think none of these will show in the UI since the names wont match any of the expected ones ('Projects', etc).

Could define an array of supported names, and then pick a random one to use:

```
const SystemAlertNames = [
  'Summary',
  'Manage Telemetry',
  'Manage Observations',
  'Manage Animals',
  'Manage Sampling',
  'Projects',
  'Surveys',
  'Standards',
  'Administrator',
  'Funding'
];

const insertSystemAlert = () => `
  ...
  $$${SystemAlertNames[Math.floor(Math.random() * SystemAlertNames.length)]}$$,
  ...
`;
```

NickPhura replied:
Disregard, misread the schema!

NickPhura commented on api/src/paths/alert/{alertId}/index.ts:
```diff
@@ -0,0 +1,310 @@
+    return {
+      and: [
+        {
+          discriminator: 'SystemUser'
```
Just checking: 

Do we expect a random user to need to fetch an individual alert? Or should this be restricted to admins?

The only consideration is that the current repo function this calls doesn't block expired alerts. Which is maybe fine? But maybe not something we want random users to be able to fetch?

If Alerts never contain sensitive information, then no issue of course.

mauberti-bc replied:
Not right now - I changed all the alert/{alertId} endpoints to be data/system admins only.

NickPhura commented on api/src/paths/alert/{alertId}/index.ts:
```diff
@@ -0,0 +1,310 @@
+  ],
+  parameters: [
+    {
+      in: 'query',
```
`in: 'path'`

and should be `required: true`

Also, should be an `'integer'`

NickPhura commented on api/src/paths/alert/{alertId}/index.ts:
```diff
@@ -0,0 +1,310 @@
+    }
+  ],
+  requestBody: {
+    description: 'Alert post request object.',
```
nit: `put`

NickPhura commented on api/src/paths/alert/{alertId}/index.ts:
```diff
@@ -0,0 +1,310 @@
+    return {
+      and: [
+        {
+          discriminator: 'SystemUser',
```
`'SystemRole'`

NickPhura commented on api/src/paths/alert/{alertId}/index.ts:
```diff
@@ -0,0 +1,310 @@
+    try {
+      await connection.open();
+
+      const alertId = Number(req.params.alertId);
+      const alert = req.body;
+
+      const alertService = new AlertService(connection);
+
+      const id = await alertService.updateAlert({ ...alert, alert_id: alertId });
```
This doesn't really matter: but since the request body already contains the alert_id, you could just use that one, rather than parsing it from the path. Though both are technically fine.

Getting the id from a specific part of the request only matters when we deal with security. But since this is an admin-only endpoint, we can trust the data we receive.

mauberti-bc replied:
Good point. I figured that the path's alertId was more reliable, in the (unlikely) scenario where the path's alertId differs from the request body's alertId.

NickPhura commented on api/src/repositories/alert-repository.ts:
```diff
@@ -0,0 +1,228 @@
+
+    const response = await this.connection.sql(sqlStatement);
+
+    const result = response?.rows?.[0] ?? null;
+
+    if (!result) {
```
Its slightly better to check `response.rowCount`, which will always be the number of rows processed. Why that matters is because if we make an update query, and forget a where clause, it might be that it updates more than 1 record.

And so we really want to check if `response.rowCount === 1`. Because less than 1 or more than 1 is equally bad (in this case).

NickPhura replied:
You'll see some checks like that in SIMS, and the message will be:
`''response.rowCount was !== 1, expected rowCount === 1''`

mauberti-bc replied:
Thanks, that makes sense

NickPhura commented on api/src/services/alert-service.ts:
```diff
@@ -0,0 +1,81 @@
+  }
+
+  /**
+   * Create and associate alert for survey.
```
This mentions 'survey' but I dont think its survey specific, same with the below functions.

NickPhura commented on api/src/services/alert-service.ts:
```diff
@@ -0,0 +1,81 @@
+   * @return {*}  Promise<number>
+   * @memberof AlertService
+   */
+  async deactivateAlert(alertId: number, recordEndDate: string): Promise<number> {
```
This is never called (outside of tests).
Same with the matching repo function.

mauberti-bc replied:
Removed - they were from an earlier iteration

NickPhura commented:
@mauberti-bc Really slick in general, nice work! I like the way the alerts are displayed in the admin page and form (with the preview).

2 things I noticed when adding/editing/deleting alerts in the UI:
- The alert name has a max length of 50 in the UI, but max length of 100 in the database. Not a major issue, but maybe they were both meant to be 50 or 100?
- When I edit an alert, if I clear the expiration date, I can't save (I get an 'Invalid date' error).

Regarding CSS:
Also, was dabbling with some CSS tweaks, if you're interested. Just on my local.

NickPhura committed:
Merge remote-tracking branch 'origin/dev' into SIMSBIOHUB-620

NickPhura committed:
ignore-skip

NickPhura committed:
CSS Tweaks

NickPhura committed:
Remove commented code

mauberti-bc commented:
> @mauberti-bc Really slick in general, nice work! I like the way the alerts are displayed in the admin page and form (with the preview).
> 
> 2 things I noticed when adding/editing/deleting alerts in the UI:
> 
>     * The alert name has a max length of 50 in the UI, but max length of 100 in the database. Not a major issue, but maybe they were both meant to be 50 or 100?
> 
>     * When I edit an alert, if I clear the expiration date, I can't save (I get an 'Invalid date' error).
> 
> 
> Regarding CSS: Also, was dabbling with some CSS tweaks, if you're interested. Just on my local.

Fixed the bug preventing end dates from being removed by updating the DateField component to correctly catch invalid dates. Also changed the max length of alert name to 50 in the database.

mauberti-bc committed:
address PR comments

NickPhura committed:
Adjust alert spacing css.

NickPhura committed:
Tweak error message

NickPhura reviewed (approved):
⚠️ 👍

NickPhura committed:
Merge branch 'dev' into SIMSBIOHUB-620

aubertinyoung commented:
> ⚠️ 👍

Is that an alert banner or a warning 🥴
